android/envoy/src/main/kotlin/org/greatfire/envoy/EnvoyTest.kt

```diff
@@ -0,0 +1,214 @@
+package org.greatfire.envoy
+
+import android.net.Uri
+import android.net.UrlQuerySanitizer
+import android.util.Log
+
+// old shadowsocks
+import android.content.Intent
+import androidx.core.content.ContextCompat
+
+// This class represents an Envoy connection: type and URL,
+// and tracks additional information, such as a URL to any
+// proxy used, e.g. Shadowsocks provides a SOCKS5 interface
+// which is what our Interceptor actually uses
+//
+// The url should be the one provided by the caller
+// testType lets us know how to treat that URL
+// some transports have an additional URL, e.g. a SOCKS5 URL
+// for a PT, that gets stored in proxyUrl
+// proxyiSEnvoy lets us know if proxyUrl is an Envoy proxy or SOCKS/HTTP
+
+data class EnvoyTest(
+    var testType: EnvoyServiceType,
+    var url: String,
+) {
+    companion object {
+        private const val TAG = "EnvoyTest"
+    }
+
+    // proxy URL for the service providing transport
+    // can be SOCKS5, HTTP(S), Envoy, see proxyIsEnvoy
+    var proxyUrl: String? = null
+    // - true means the proxyUrl refers to a nonstandard Envoy proxy
+    // as documented at http://github.com/greatfire/envoy/
+    // - false means it's SOCKS or HTTP
+    var proxyIsEnvoy: Boolean = false
+
+    // Is this the service we chose to use
+    var selectedService: Boolean = false
+
+    // Is the associated service (if any) running?
+    var serviceRunning: Boolean = false
+
+    // envoy:// URL related options
+    val headers = mutableListOf<Pair<String, String>>()
+    // address param creates an resolver rule
+    // stash it here in case we can support it with OkHttp
+    var address: String? = null
+    var resolverRules: String? = null
+
+    // Envoy Global settings and state
+    private val state = EnvoyState.getInstance()
+
+    // used to time how long it takes to connect and test
+    private var timer: Timer? = null
+    // should this be in settings?
+    // private var shadowsocks: EnvoyShadowsocks? = null
+    private val shadowsocksIntent: Intent? = null
+
+    override fun toString(): String {
+        return UrlUtil.sanitizeUrl(url) + " (" + testType + ")"
+    }
+
+    private fun getTimer(): Timer {
+        if (timer == null) {
+            timer = Timer()
+        }
+        return timer!!
+    }
+
+    // helper to time things
+    inner class Timer() {
+        private val startTime = System.currentTimeMillis()
+        private var stopTime: Long? = null
+
+        fun stop(): Long {
+            stopTime = System.currentTimeMillis()
+            return stopTime!! - startTime
+        }
+
+        fun timeSpent(): Long {
+            if (stopTime == null) {
+                Log.e(TAG, "timeSpent called before stop()!")
+                return 0
```

Comment:
would it make sense to return the difference between the current time and the start time instead?